add +kaoriya marks

diff --git a/src/evalfunc.c b/src/evalfunc.c
index 3af865375..9a8b25120 100644
--- a/src/evalfunc.c
+++ b/src/evalfunc.c
@@ -5836,6 +5836,7 @@ f_has(typval_T *argvars, typval_T *rettv)
 #endif
 		},
 	{"jumplist", 1},
+	{"kaoriya", 1},
 	{"keymap",
 #ifdef FEAT_KEYMAP
 		1
diff --git a/src/version.c b/src/version.c
index 691a4880f..e629494ab 100644
--- a/src/version.c
+++ b/src/version.c
@@ -287,6 +287,7 @@ static char *(features[]) =
 	"-job",
 #endif
 	"+jumplist",
+	"+kaoriya",
 #ifdef FEAT_KEYMAP
 	"+keymap",
 #else
