diff --git a/src/gui.c b/src/gui.c
index e4af17a3c..89f2e9918 100644
--- a/src/gui.c
+++ b/src/gui.c
@@ -1135,7 +1135,10 @@ gui_update_cursor(
 
 		if (
 # if defined(FEAT_GUI_GTK) && defined(FEAT_XIM) && !defined(FEAT_HANGULIN)
+#  if 0
 			preedit_get_status()
+#  endif
+			im_get_status()
 # else
 			im_get_status()
 # endif
diff --git a/src/mbyte.c b/src/mbyte.c
index dc7efa96f..2c831b793 100644
--- a/src/mbyte.c
+++ b/src/mbyte.c
@@ -4793,10 +4793,7 @@ iconv_end(void)
 # endif /* USE_ICONV */
 
 
-#ifdef FEAT_GUI
-# define USE_IMACTIVATEFUNC (!gui.in_use && *p_imaf != NUL)
-# define USE_IMSTATUSFUNC (!gui.in_use && *p_imsf != NUL)
-#else
+#ifdef FEAT_EVAL
 # define USE_IMACTIVATEFUNC (*p_imaf != NUL)
 # define USE_IMSTATUSFUNC (*p_imsf != NUL)
 #endif
diff --git a/src/ui.c b/src/ui.c
index 5c9077c05..0e844dc53 100644
--- a/src/ui.c
+++ b/src/ui.c
@@ -3515,12 +3515,14 @@ im_save_status(long *psave)
      * And don't save when the GUI is running but our window doesn't have
      * input focus (e.g., when a find dialog is open). */
     if (!p_imdisable && KeyTyped && !KeyStuffed
+#if 0
 # ifdef FEAT_XIM
 	    && xic != NULL
 # endif
 # ifdef FEAT_GUI
 	    && (!gui.in_use || gui.in_focus)
 # endif
+#endif
 	)
     {
 	/* Do save when IM is on, or IM is off and saved status is on. */
