implement 'transparency' for Windows GUI

diff --git a/src/gui_w32.c b/src/gui_w32.c
index 6e12b9458..695791ee4 100644
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -50,6 +50,14 @@ static int gui_mswin_get_menu_height(int fix_window);
 # define gui_mswin_get_menu_height(fix_window)	0
 #endif
 
+//
+// For Transparent Window.
+//
+typedef DWORD (WINAPI *FWINLAYER)(HWND hwnd, DWORD crKey, BYTE bAlpha,
+	DWORD dwFlags);
+static void w32_set_transparency(HWND hwnd, BYTE bAlpha);
+
+
 typedef struct keycode_trans_strategy {
     void (*ptr_on_char) (HWND /*hwnd UNUSED*/, UINT /*cch*/, int /*cRepeat UNUSED*/);
     void (*ptr_on_sys_char) (HWND /*hwnd UNUSED*/, UINT /*cch*/, int /*cRepeat UNUSED*/);
@@ -6004,6 +6012,8 @@ gui_mch_init(void)
 	(void)gui_mch_set_rendering_options(p_rop);
 #endif
 
+    w32_set_transparency(s_hwnd, p_transparency);
+
 theend:
     // Display any pending error messages
     display_errors();
@@ -6192,6 +6202,44 @@ gui_mch_set_sp_color(guicolor_T color)
     gui.currSpColor = color;
 }
 
+    void
+w32_set_transparency(HWND hwnd, BYTE bAlpha)
+{
+    FWINLAYER pfLayer;
+    HANDLE hDll;
+
+    if (!hwnd)
+	hwnd = s_hwnd;
+
+    // Turn off transpareny
+    if (bAlpha == 255)
+    {
+	SetWindowLong(hwnd, GWL_EXSTYLE, ~WS_EX_LAYERED &
+		GetWindowLong(hwnd, GWL_EXSTYLE));
+	return;
+    }
+
+    // Obtain pointer to function set transparecy rate
+    if (!(hDll = LoadLibrary("user32.dll")))
+	return;
+    pfLayer = (FWINLAYER)GetProcAddress(hDll, "SetLayeredWindowAttributes");
+
+    if (pfLayer)
+    {
+	SetWindowLong(hwnd, GWL_EXSTYLE, WS_EX_LAYERED |
+		GetWindowLong(hwnd, GWL_EXSTYLE));
+	pfLayer(hwnd, 0, bAlpha, LWA_ALPHA);
+    }
+
+    FreeLibrary(hDll);
+}
+
+    void
+gui_mch_set_transparency(int alpha)
+{
+    w32_set_transparency(NULL, (BYTE)alpha);
+}
+
 #ifdef FEAT_MBYTE_IME
 /*
  * Multi-byte handling, originally by Sung-Hoon Baek.
diff --git a/src/option.c b/src/option.c
index 1ec7efbf8..7f66cbd1a 100644
--- a/src/option.c
+++ b/src/option.c
@@ -4542,6 +4542,21 @@ did_set_titlelen(optset_T *args)
     return errmsg;
 }
 
+/*
+ * Process the new 'transparency' option value.
+ */
+    char *
+did_set_transparency(optset_T *args UNUSED)
+{
+    if (p_transparency < 1 || p_transparency > 255)
+	p_transparency = 255;
+# ifdef VIMDLL
+    if (gui.in_use)
+# endif
+	gui_mch_set_transparency(p_transparency);
+    return NULL;
+}
+
 #if defined(FEAT_PERSISTENT_UNDO)
 /*
  * Process the updated 'undofile' option value.
diff --git a/src/option.h b/src/option.h
index 7372ba924..67ebdca81 100644
--- a/src/option.h
+++ b/src/option.h
@@ -1051,6 +1051,7 @@ EXTERN long	p_titlelen;	// 'titlelen'
 EXTERN char_u	*p_titleold;	// 'titleold'
 EXTERN char_u	*p_titlestring;	// 'titlestring'
 EXTERN char_u	*p_tsr;		// 'thesaurus'
+EXTERN long	p_transparency;	// 'transparency'
 EXTERN int	p_ttimeout;	// 'ttimeout'
 EXTERN long	p_ttm;		// 'ttimeoutlen'
 EXTERN int	p_tbi;		// 'ttybuiltin'
diff --git a/src/optiondefs.h b/src/optiondefs.h
index 9e5612470..29f7a4096 100644
--- a/src/optiondefs.h
+++ b/src/optiondefs.h
@@ -2763,6 +2763,11 @@ static struct vimoption options[] =
 			    {(char_u *)0L, (char_u *)0L}
 #endif
 			    SCTX_INIT},
+    {"transparency", "tra", P_NUM|P_VI_DEF,
+			    (char_u *)&p_transparency, PV_NONE,
+			    did_set_transparency, NULL,
+			    {(char_u *)255L, (char_u *)0L}
+			    SCTX_INIT},
     {"ttimeout",    NULL,   P_BOOL|P_VI_DEF|P_VIM,
 			    (char_u *)&p_ttimeout, PV_NONE, NULL, NULL,
 			    {(char_u *)FALSE, (char_u *)0L} SCTX_INIT},
diff --git a/src/proto/gui_w32.pro b/src/proto/gui_w32.pro
index a93df3672..acb87f1c5 100644
--- a/src/proto/gui_w32.pro
+++ b/src/proto/gui_w32.pro
@@ -74,6 +74,7 @@ void gui_mch_set_font(GuiFont font);
 void gui_mch_set_fg_color(guicolor_T color);
 void gui_mch_set_bg_color(guicolor_T color);
 void gui_mch_set_sp_color(guicolor_T color);
+void gui_mch_set_transparency(int alpha);
 void im_set_font(LOGFONTW *lf);
 void im_set_position(int row, int col);
 void im_set_active(int active);
diff --git a/src/proto/option.pro b/src/proto/option.pro
index 4fa8935c0..cb4486914 100644
--- a/src/proto/option.pro
+++ b/src/proto/option.pro
@@ -79,6 +79,7 @@ char *did_set_textmode(optset_T *args);
 char *did_set_textwidth(optset_T *args);
 char *did_set_title_icon(optset_T *args);
 char *did_set_titlelen(optset_T *args);
+char *did_set_transparency(optset_T *args);
 char *did_set_undofile(optset_T *args);
 char *did_set_undolevels(optset_T *args);
 char *did_set_updatecount(optset_T *args);
