implement 'charspace' for Windows GUI

diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
index b1550dfcc..e3039e70b 100644
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -1760,6 +1760,20 @@ A jump table for the options with a short description can be found at |Q_op|.
 	(the most recently created) will be used in its place.  For additional
 	info, see |quickfix-stack|.
 
+						*'charspace'* *'csp'*
+'charspace' 'csp'	number (default: 0)
+			global
+			{not in Vi}
+			{only in Windows GUI version and when compiled with
+			the |+kaoriya| feature}
+	Number of horizontal pixels inserted between characters.  Useful if
+	the font uses the full character cell width, making characters touch
+	each other.
+	When non-zero there is room for side of characters.
+	With some fonts there can  be too much room between characters (to
+	have space for left and right).  Then it make sense to set 'charspace'
+	to a negative value.  This may cause display problems though!
+
 				   *'cindent'* *'cin'* *'nocindent'* *'nocin'*
 'cindent' 'cin'		boolean	(default off)
 			local to buffer
diff --git a/src/gui_w32.c b/src/gui_w32.c
index 695791ee4..f04709e28 100644
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -1861,7 +1861,11 @@ GetFontSize(GuiFont font, int *char_width, int *char_height)
     GetAverageFontSize(hdc, &size);
 
     if (char_width)
+    {
 	*char_width = size.cx + tm.tmOverhang;
+	// add extra pixels for 'charspace' option.
+	*char_width += p_charspace;
+    }
     if (char_height)
 	*char_height = tm.tmHeight + p_linespace;
 
diff --git a/src/option.h b/src/option.h
index ff727a4ee..bcc3637f1 100644
--- a/src/option.h
+++ b/src/option.h
@@ -503,6 +503,9 @@ EXTERN int	p_deco;		// 'delcombine'
 EXTERN char_u	*p_ccv;		// 'charconvert'
 #endif
 EXTERN int	p_cdh;		// 'cdhome'
+#ifdef FEAT_GUI
+EXTERN long	p_charspace;	// 'charspace'
+#endif
 EXTERN char_u	*p_cino;	// 'cinoptions'
 EXTERN char_u	*p_cedit;	// 'cedit'
 EXTERN long	p_cwh;		// 'cmdwinheight'
diff --git a/src/optiondefs.h b/src/optiondefs.h
index 7d2ece2d1..f19382384 100644
--- a/src/optiondefs.h
+++ b/src/optiondefs.h
@@ -590,6 +590,14 @@ static struct vimoption options[] =
 			    {(char_u *)0L, (char_u *)0L}
 #endif
 			    SCTX_INIT},
+    {"charspace",   "csp",  P_NUM|P_NODEFAULT|P_VIM|P_RCLR,
+#ifdef FEAT_GUI
+			    (char_u *)&p_charspace, PV_NONE, did_set_linespace,
+#else
+			    (char_u *)NULL, PV_NONE, NULL,
+#endif
+			    NULL, {(char_u *)0L, (char_u *)0L}
+			    SCTX_INIT},
     {"chistory",    "chi",  P_NUM|P_VI_DEF,
 #ifdef FEAT_QUICKFIX
 			    (char_u *)&p_chi, PV_NONE, did_set_xhistory, NULL,
