add 'auto' for 'ambiwidth'

diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
index b00dacf4b..d5c79fd04 100644
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -817,6 +817,9 @@ A jump table for the options with a short description can be found at |Q_op|.
 	"single":	Use the same width as characters in US-ASCII.  This is
 			expected by most users.
 	"double":	Use twice the width of ASCII characters.
+	"auto":		Use obtained width from system.
+			{only for Win NT/2K systems, currently}
+
 							*E834* *E835*
 	The value "double" cannot be used if 'listchars' or 'fillchars'
 	contains a character that would be double width.  These errors may
@@ -845,6 +848,10 @@ A jump table for the options with a short description can be found at |Q_op|.
 	escape sequence to request cursor position report.  The response can
 	be found in |v:termu7resp|.
 
+	On some systems, we can use width for a character which defined by
+	font design.  If you use such a system try "auto" for this option.
+	It will give you more nice result for a font which you select.
+
 			*'antialias'* *'anti'* *'noantialias'* *'noanti'*
 'antialias' 'anti'	boolean (default: off)
 			global
diff --git a/src/gui_w32.c b/src/gui_w32.c
index 7765eccbd..c1d7ce92a 100644
--- a/src/gui_w32.c
+++ b/src/gui_w32.c
@@ -5670,6 +5670,56 @@ error:
 }
 #endif
 
+#ifdef USE_AMBIWIDTH_AUTO
+#define CHARWIDE_CACHESIZE 0x20000
+static GuiFont last_font = NOFONT;
+
+    int
+gui_mch_get_charwidth(int c)
+{
+    static int cache[CHARWIDE_CACHESIZE];
+    GuiFont usingfont = gui.wide_font ? gui.wide_font : gui.norm_font;
+
+    // Check validity of charwide cache
+    if (last_font != usingfont)
+    {
+	// Update cache. -1 is mark for uninitialized cell
+	int i;
+
+	TRACE("Charwide cache will be updated (base=%d)\n", gui.char_width);
+	last_font = usingfont;
+	for (i = 0; i < CHARWIDE_CACHESIZE; ++i)
+	    cache[i] = -1;
+    }
+    if (usingfont && 0 <= c && c < CHARWIDE_CACHESIZE)
+    {
+	if (cache[c] >= 0)
+	    return cache[c]; // Use cached value
+	else
+	{
+	    // Get true character width in dot, convert to cells and save it.
+	    int	    len;
+	    ABC	    fontABC;
+	    HFONT   hfntOld = SelectFont(s_hdc, usingfont);
+
+	    if (!GetCharABCWidthsW(s_hdc, c, c, &fontABC) ||
+		    (len = fontABC.abcA + fontABC.abcB + fontABC.abcC) <= 0)
+	    {
+		TRACE("GetCharABCWidthsW() failed for %08X\n", c);
+		cache[c] = 0;
+	    }
+	    else
+		cache[c] = ((len + (gui.char_width >> 1)) / gui.char_width);
+	    SelectFont(s_hdc, hfntOld);
+
+	    return cache[c];
+	}
+    }
+    else
+	return 0;
+}
+#endif
+
 /*
  * Parse the GUI related command-line arguments.  Any arguments used are
  * deleted from argv, and *argc is decremented accordingly.  This is called
diff --git a/src/mbyte.c b/src/mbyte.c
index 38d0fe5a8..c602f5c83 100644
--- a/src/mbyte.c
+++ b/src/mbyte.c
@@ -1593,6 +1593,19 @@ utf_char2cells(int c)
     }
 #endif
 
+#ifdef USE_AMBIWIDTH_AUTO
+    if (gui.in_use && *p_ambw == 'a')
+    {
+	int cell;
+
+	/* This is required by screen.c implicitly. */
+	if (c == 0)
+	    return 1;
+	if (utf_printable(c) && (cell = gui_mch_get_charwidth(c)) > 0)
+	    return cell;
+    }
+#endif
+
     if (c >= 0x100)
     {
 	if (!utf_printable(c))
diff --git a/src/optionstr.c b/src/optionstr.c
index 84f973442..24444a53f 100644
--- a/src/optionstr.c
+++ b/src/optionstr.c
@@ -16,7 +16,11 @@
 static char_u shm_buf[SHM_LEN];
 static int set_shm_recursive = 0;
 
-static char *(p_ambw_values[]) = {"single", "double", NULL};
+static char *(p_ambw_values[]) = {"single", "double",
+# ifdef USE_AMBIWIDTH_AUTO
+    "auto",
+# endif
+    NULL};
 static char *(p_bg_values[]) = {"light", "dark", NULL};
 static char *(p_bkc_values[]) = {"yes", "auto", "no", "breaksymlink", "breakhardlink", NULL};
 static char *(p_bo_values[]) = {"all", "backspace", "cursor", "complete",
diff --git a/src/proto/gui_w32.pro b/src/proto/gui_w32.pro
index 89443b0f6..a348ed94d 100644
--- a/src/proto/gui_w32.pro
+++ b/src/proto/gui_w32.pro
@@ -82,6 +82,7 @@ void im_set_active(int active);
 int im_get_status(void);
 void gui_mch_draw_string(int row, int col, char_u *text, int len, int flags);
 void gui_mch_flush(void);
+int gui_mch_get_charwidth(int c);
 void gui_mch_get_screen_dimensions(int *screen_w, int *screen_h);
 void gui_mch_add_menu(vimmenu_T *menu, int pos);
 void gui_mch_show_popupmenu(vimmenu_T *menu);
diff --git a/src/vim.h b/src/vim.h
index 4bd6c7001..daf3c5971 100644
--- a/src/vim.h
+++ b/src/vim.h
@@ -576,6 +576,13 @@ typedef long long vimlong_T;
 # endif
 #endif
 
+/*
+ * Whether 'ambiwidth' supports "auto".  Currently, only for Win32.
+ */
+#if defined(FEAT_GUI) && defined(FEAT_GUI_MSWIN)
+# define USE_AMBIWIDTH_AUTO
+#endif
+
 /*
  * For dynamically loaded gettext library.  Currently, only for Win32.
  */
