support 'imaf' and 'imsf' on GTK

diff --git a/src/gui.c b/src/gui.c
index 29f16effd..552557a96 100644
--- a/src/gui.c
+++ b/src/gui.c
@@ -1243,7 +1243,7 @@ gui_update_cursor(
 		guicolor_T fg, bg;
 
 		if (
-# if defined(FEAT_GUI_GTK) && defined(FEAT_XIM)
+# if 0 && defined(FEAT_GUI_GTK) && defined(FEAT_XIM)
 			preedit_get_status()
 # else
 			im_get_status()
diff --git a/src/ui.c b/src/ui.c
index b56e39525..518c0a562 100644
--- a/src/ui.c
+++ b/src/ui.c
@@ -1184,12 +1184,14 @@ im_save_status(long *psave)
     // And don't save when the GUI is running but our window doesn't have
     // input focus (e.g., when a find dialog is open).
     if (!p_imdisable && KeyTyped && !KeyStuffed
+#if 0
 # ifdef FEAT_XIM
 	    && xic != NULL
 # endif
 # ifdef FEAT_GUI
 	    && (!gui.in_use || gui.in_focus)
 # endif
+#endif
 	)
     {
 	// Do save when IM is on, or IM is off and saved status is on.
