make 2html.vim auto link URL which starts 'ftp'

diff --git a/runtime/syntax/2html.vim b/runtime/syntax/2html.vim
index 527e7d38b..039c7dbe2 100644
--- a/runtime/syntax/2html.vim
+++ b/runtime/syntax/2html.vim
@@ -2006,7 +2006,7 @@ endif
 
 # Add hyperlinks
 if !settings.no_links
-  :%s+\(https\=://\S\{-}\)\(\([.,;:}]\=\(\s\|$\)\)\|[\\"'<>]\|&gt;\|&lt;\|&quot;\)+<a href="\1">\1</a>\2+ge
+  :%s+\(\%(https\=\|ftp\)://\S\{-}\)\(\([.,;:}]\=\(\s\|$\)\)\|[\\"'<>]\|&gt;\|&lt;\|&quot;\)+<a href="\1">\1</a>\2+ge
 endif
 
 # The DTD
