add +kaoriya marks

diff --git a/src/evalfunc.c b/src/evalfunc.c
index 52c7897a9..b799a96b6 100644
--- a/src/evalfunc.c
+++ b/src/evalfunc.c
@@ -7182,6 +7182,7 @@ f_has(typval_T *argvars, typval_T *rettv)
 #endif
 		},
 	{"jumplist", 1},
+	{"kaoriya", 1},
 	{"keymap",
 #ifdef FEAT_KEYMAP
 		1
diff --git a/src/version.c b/src/version.c
index 6da786909..216e445ad 100644
--- a/src/version.c
+++ b/src/version.c
@@ -288,6 +288,7 @@ static char *(features[]) =
 	"-job",
 #endif
 	"+jumplist",
+	"+kaoriya",
 #ifdef FEAT_KEYMAP
 	"+keymap",
 #else
