add +kaoriya marks

diff --git a/src/evalfunc.c b/src/evalfunc.c
index 00da8d881..2a1835eda 100644
--- a/src/evalfunc.c
+++ b/src/evalfunc.c
@@ -7198,6 +7198,7 @@ f_has(typval_T *argvars, typval_T *rettv)
 #endif
 		},
 	{"jumplist", 1},
+	{"kaoriya", 1},
 	{"keymap",
 #ifdef FEAT_KEYMAP
 		1
diff --git a/src/version.c b/src/version.c
index f6d341d76..dd9ffc666 100644
--- a/src/version.c
+++ b/src/version.c
@@ -288,6 +288,7 @@ static char *(features[]) =
 	"-job",
 #endif
 	"+jumplist",
+	"+kaoriya",
 #ifdef FEAT_KEYMAP
 	"+keymap",
 #else
