add +kaoriya marks

diff --git a/src/evalfunc.c b/src/evalfunc.c
index b896e3fdd..11975cd25 100644
--- a/src/evalfunc.c
+++ b/src/evalfunc.c
@@ -7327,6 +7327,7 @@ f_has(typval_T *argvars, typval_T *rettv)
 #endif
 		},
 	{"jumplist", 1},
+	{"kaoriya", 1},
 	{"keymap",
 #ifdef FEAT_KEYMAP
 		1
diff --git a/src/version.c b/src/version.c
index c69bfc961..45a79378c 100644
--- a/src/version.c
+++ b/src/version.c
@@ -318,6 +318,7 @@ static char *(features[]) =
 	"-job",
 #endif
 	"+jumplist",
+	"+kaoriya",
 #ifdef FEAT_KEYMAP
 	"+keymap",
 #else
