# HG changeset patch
# Parent  b82bd3dbc798fac5f8ac73e4c362978a3874c9ef

diff --git a/src/getchar.c b/src/getchar.c
--- a/src/getchar.c
+++ b/src/getchar.c
@@ -3138,11 +3138,12 @@ fix_input_buffer(char_u *buf, int len)
 		    /* timeout may generate K_CURSORHOLD */
 		    && (i < 2 || p[1] != KS_EXTRA || p[2] != (int)KE_CURSORHOLD)
 #if defined(MSWIN) && (!defined(FEAT_GUI) || defined(VIMDLL))
+		    /* Win32 console passes modifiers */
+		    && (
 # ifdef VIMDLL
-		    && !gui.in_use
+			gui.in_use ? 1 :
 # endif
-		    /* Win32 console passes modifiers */
-		    && (i < 2 || p[1] != KS_MODIFIER)
+			(i < 2 || p[1] != KS_MODIFIER))
 #endif
 		    ))
 	{
