# HG changeset patch
# Parent  0066c86671c13627d64536202cdea8184ec26654

diff -r 0066c86671c1 -r fe7c4cc6dddf src/os_win32.c
--- a/src/os_win32.c	Thu Mar 26 21:09:06 2015 +0900
+++ b/src/os_win32.c	Sat Mar 28 21:32:39 2015 +0900
@@ -5398,6 +5398,13 @@
     COORD coord = g_coord;
     DWORD written;
 
+    _RPT2(_CRT_WARN, "%d, %d", coord.X, coord.Y);
+    _RPT3(_CRT_WARN, ", %d : %.*s\n", cchToWrite, cchToWrite, pchBuf);
+#if 1
+    if (coord.X == 4 && coord.Y == 0 && cchToWrite == 2) {
+	_RPT0(_CRT_WARN, "***HERE***\n");
+    }
+#endif
     FillConsoleOutputAttribute(g_hConOut, g_attrCurrent, cchToWrite,
 				coord, &written);
     /* When writing fails or didn't write a single character, pretend one
